op-gateway/src/OPProofService.ts

```diff
@@ -15,58 +42,200 @@ const L2_OUTPUT_ORACLE_ABI = [
 const L2_TO_L1_MESSAGE_PASSER_ADDRESS =
   '0x4200000000000000000000000000000000000016';
 
+const DISPUTE_GAME_FACTORY_ABI = [
+  'function gameAtIndex(uint256 _index) external view returns (uint32 gameType_, uint64 timestamp_, address proxy_)',
+  'function gameCount() external view returns (uint256 gameCount_)',
+  'function findLatestGames(uint32 _gameType, uint256 _start, uint256 _n) external view returns (tuple(uint256 index, bytes32 metadata, uint64 timestamp, bytes32 rootClaim, bytes extraData)[] memory games_)',
+];
+
+const OPTIMISM_PORTAL_ABI = [
+  'function l2Oracle() external view returns (address)',
+  'function disputeGameFactory() external view returns (address)',
+  'function respectedGameType() external view returns (uint32)',
+];
+
+const FAULT_DISPUTE_GAME_ABI = [
+  // The l2BlockNumber of the disputed output root in the `L2OutputOracle`.
+  'function l2BlockNumber() external view returns (uint256 l2BlockNumber_)',
+  // The output root of the game
+  'function rootClaim() external view returns (bytes32 rootClaim_)',
+  // Status of the game challenging
+  'function status() external view returns (uint8)',
+];
+
 /**
  * The proofService class can be used to calculate proofs for a given target and slot on the Optimism Bedrock network.
  * It's also capable of proofing long types such as mappings or string by using all included slots in the proof.
  *
  */
 export class OPProofService implements IProofService<OPProvableBlock> {
-  private readonly l2OutputOracle: Contract;
+  private readonly l1Provider: JsonRpcProvider;
   private readonly l2Provider: JsonRpcProvider;
   private readonly helper: EVMProofHelper;
   private readonly delay: number;
 
+  private readonly optimismPortal: Contract;
+  private l2OutputOracle: Contract | undefined;
+  private disputeGameFactory: Contract | undefined;
+
   constructor(
     l1Provider: JsonRpcProvider,
     l2Provider: JsonRpcProvider,
-    l2OutputOracleAddress: string,
+    optimismPortalAddress: string,
     delay: number
   ) {
+    this.l1Provider = l1Provider;
     this.l2Provider = l2Provider;
     this.helper = new EVMProofHelper(l2Provider);
     this.delay = delay;
-    this.l2OutputOracle = new Contract(
-      l2OutputOracleAddress,
-      L2_OUTPUT_ORACLE_ABI,
+
+    this.optimismPortal = new Contract(
+      optimismPortalAddress,
+      OPTIMISM_PORTAL_ABI,
       l1Provider
     );
   }
 
+  async findDisputeGame(): Promise<OPProvableBlock | null> {
+    if (!this.disputeGameFactory) {
+      try {
+        const disputeGameFactoryAddress = await this.optimismPortal.disputeGameFactory();
+  
+        this.disputeGameFactory = new Contract(
+          disputeGameFactoryAddress,
+          DISPUTE_GAME_FACTORY_ABI,
+          this.l1Provider
+        );
+      } catch (err) {
```

Comment:
An error setting this address, or other issue, should be cached here to avoid trying to do it on every call. Perhaps this and the other contract could both be initialized in the constructor instead of doing it just-in-time?

Reply:
Using OPOutputLookup contract I have built at https://github.com/Opti-domains/dispute-game-lookup/blob/main/src/OPOutputLookup.sol to do the calculation and advanced search in a single call. So, the lowest latency one can acheive.